example/extensions/lib_subgraph/subgraph_lib.cc

```diff
@@ -0,0 +1,278 @@
+    }
+  }
+
+  // free allocated temporary storage
+  for (void* ptr : to_free) {
+    free(ptr);
+  }
+  
+  return MX_SUCCESS;
+}
+
+class MyStatefulOp : public CustomStatefulOp {
+ public:
+  explicit MyStatefulOp(std::string sym) : subgraph_sym(sym) {}
+
+  MXReturnValue Forward(std::vector<MXTensor> inputs,
+                        std::vector<MXTensor> outputs,
+                        OpResource op_res) {
+    return myExecutor(inputs, outputs, subgraph_sym);
+  }
+
+ private:
+  std::string subgraph_sym;
+};
+
+MXReturnValue createOpState(std::map<std::string, std::string> attrs,
+                            CustomStatefulOp** op_inst) {
+  std::string serialized_subgraph = "[empty]";
+  // MXNet subgraph is stored as Symbol in operator node attrs subgraphs field
+  // custom subgraph is stored as json string in custom operator attrs map entry
+  if (attrs.count(SUBGRAPH_SYM_JSON)) {
+    // user can now parse json and run other custom ops inside subgraph
+    serialized_subgraph = attrs[SUBGRAPH_SYM_JSON];
+  }
+  *op_inst = new MyStatefulOp(serialized_subgraph);
+  std::cout << "Info: stateful operator created" << std::endl;
+  return MX_SUCCESS;
+}
+
+REGISTER_OP(_custom_subgraph_op)
+.setParseAttrs(parseAttrs)
+.setInferType(inferType)
+.setInferShape(inferShape)
+.setCreateOpState(createOpState);
+
+const std::vector<std::string> op_names({"exp","log"});
+
+MXReturnValue mySupportedOps(std::string json,
+                             const int num_ids,
+                             int *ids,
+                             std::map<std::string, std::string> options) {
+  for (auto kv : options) {
+    std::cout << "option: " << kv.first << " ==> " << kv.second << std::endl;
+  }
+  //convert json string to json object
+  JsonParser parser;
+  JsonVal json_val = parser.parse_to_json(json);
+  //get nodes list
+  JsonVal nodes = json_val.map[JsonVal("nodes")];
+
+  //loop over nodes
+  for(int i=0; i<nodes.list.size(); i++) {
+    JsonVal node = nodes.list[i];
+    JsonVal op = node.map[JsonVal("op")];
+
+    //get shape/type if available
+    std::string shape;
+    int dtype = -1;
+    if(node.map.find(JsonVal("attrs")) != node.map.end()) {
+      JsonVal attrs = node.map[JsonVal("attrs")];
+      if(attrs.map.find(JsonVal("shape")) != attrs.map.end()) 
+        shape = attrs.map[JsonVal("shape")].str;
+      if(attrs.map.find(JsonVal("dtype")) != attrs.map.end())
+        dtype = std::stoi(attrs.map[JsonVal("dtype")].str);
+    }
+
+    //check if op dtype is float
+    if(dtype == kFloat32) {
+      //check if op is in whitelist
+      if(std::find(op_names.begin(),op_names.end(),op.str.c_str()) != op_names.end()) {
+        // found op in whitelist, set value to 1 to include op in subgraph
+        ids[i]=1;
+      }
+    }
+  }
+  return MX_SUCCESS;
+}
+
+MXReturnValue myAcceptSubgraph(std::string json, int subraph_id, bool* accept,
+                               std::map<std::string, std::string> options) {
```

Comment:
Should we keep the cout here?

Reply:
no, we can remove it. doesnt see to make any useful difference

Reply:
nah, ill take it out